File: public_web/src/lib/vixrexBlogIntent.ts
export type VixrexBlogIntent =
  | "blog_yazisi_bul"
  | "blog_yazisi_oner"
  | "blog_yazisi_taslak_ekle"
  | "blog_taslaklarini_listele"
  | "blog_taslak_duzenle"
  | "blog_yayinla";

export type VixrexAssistantDomainDecision =
  | { domain: "storefront"; intent: null; query: "" }
  | { domain: "blog"; intent: VixrexBlogIntent; query: string }
  | { domain: "ambiguous"; intent: null; query: string };

function normalize(input: string): string {
  return input
    .toLocaleLowerCase("tr-TR")
    .replace(/[ıİ]/g, "i")
    .replace(/[şŞ]/g, "s")
    .replace(/[ğĞ]/g, "g")
    .replace(/[üÜ]/g, "u")
    .replace(/[öÖ]/g, "o")
    .replace(/[çÇ]/g, "c")
    .replace(/[^a-z0-9]+/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function hasAny(text: string, tokens: readonly string[]): boolean {
  return tokens.some((token) => text.includes(token));
}

function queryFrom(text: string): string {
  const stop = new Set([
    "blog",
    "bloga",
    "blogda",
    "blogdan",
    "bloguna",
    "blogunda",
    "vitrin",
    "vitrinin",
    "vitrine",
    "vitrinime",
    "yazi",
    "yazisi",
    "yazisini",
    "yazilar",
    "yazilarimi",
    "makale",
    "rehber",
    "taslak",
    "taslagi",
    "taslagini",
    "taslaklar",
    "taslaklarimi",
    "ekle",
    "aktar",
    "cek",
    "olustur",
    "bul",
    "ara",
    "goster",
    "listele",
    "oner",
    "duzenle",
    "degistir",
    "yayinla",
    "yayina",
    "al",
    "bana",
    "bir",
  ]);
  return text
    .split(" ")
    .filter((token) => token.length > 1 && !stop.has(token))
    .join(" ")
    .trim();
}

/**
 * Üst seviye domain router'ın Blog tarafı.
 *
 * Kritik sınır: yalnız "blog" kelimesi Blog domain'e geçirmez.
 * Örn. "blog başlığını değiştir" 46-alan storefront motorunda kalır.
 * Blog domain yalnız açık içerik komutlarıyla seçilir.
 */
export function routeVixrexAssistantDomain(input: string): VixrexAssistantDomainDecision {
  const text = normalize(input);
  if (!text) return { domain: "storefront", intent: null, query: "" };

  const mentionsBlog = text.includes("blog");
  const mentionsContent = hasAny(text, ["yazi", "makale", "rehber", "taslak", "taslag"]);
  if (!mentionsBlog || !mentionsContent) {
    return { domain: "storefront", intent: null, query: "" };
  }

  const matches: VixrexBlogIntent[] = [];

  if (
    hasAny(text, ["taslak", "taslag"]) &&
    hasAny(text, ["listele", "goster", "neler", "hangileri"])
  ) {
    matches.push("blog_taslaklarini_listele");
  }
  if (hasAny(text, ["yayinla", "yayina al"])) {
    matches.push("blog_yayinla");
  }
  if (
    hasAny(text, ["duzenle", "degistir"]) &&
    hasAny(text, ["yazi", "taslak", "taslag", "makale"])
  ) {
    matches.push("blog_taslak_duzenle");
  }
  if (
    hasAny(text, ["ekle", "aktar", "cek", "taslak olustur", "taslak ekle"]) &&
    hasAny(text, ["yazi", "makale", "rehber"])
  ) {
    matches.push("blog_yazisi_taslak_ekle");
  }
  if (text.includes("oner")) {
    matches.push("blog_yazisi_oner");
  }
  if (hasAny(text, ["bul", "ara"])) {
    matches.push("blog_yazisi_bul");
  }

  const unique = [...new Set(matches)];
  const query = queryFrom(text);
  if (unique.length > 1) return { domain: "ambiguous", intent: null, query };
  if (unique.length === 1) return { domain: "blog", intent: unique[0], query };
  return { domain: "storefront", intent: null, query: "" };
}

File: lib/services/vixrex_blog_intent.dart
enum VixrexBlogIntent {
  blogYazisiBul,
  blogYazisiOner,
  blogYazisiTaslakEkle,
  blogTaslaklariniListele,
  blogTaslakDuzenle,
  blogYayinla,
}

enum VixrexAssistantDomain { storefront, blog, ambiguous }

class VixrexAssistantDomainDecision {
  final VixrexAssistantDomain domain;
  final VixrexBlogIntent? intent;
  final String query;

  const VixrexAssistantDomainDecision({
    required this.domain,
    this.intent,
    this.query = '',
  });
}

String _normalize(String input) {
  return input
      .toLowerCase()
      .replaceAll('ı', 'i')
      .replaceAll('ş', 's')
      .replaceAll('ğ', 'g')
      .replaceAll('ü', 'u')
      .replaceAll('ö', 'o')
      .replaceAll('ç', 'c')
      .replaceAll(RegExp(r'[^a-z0-9]+'), ' ')
      .replaceAll(RegExp(r'\s+'), ' ')
      .trim();
}

bool _hasAny(String text, List<String> tokens) =>
    tokens.any((token) => text.contains(token));

String _queryFrom(String text) {
  const stop = <String>{
    'blog',
    'bloga',
    'blogda',
    'blogdan',
    'bloguna',
    'blogunda',
    'vitrin',
    'vitrinin',
    'vitrine',
    'vitrinime',
    'yazi',
    'yazisi',
    'yazisini',
    'yazilar',
    'yazilarimi',
    'makale',
    'rehber',
    'taslak',
    'taslagi',
    'taslagini',
    'taslaklar',
    'taslaklarimi',
    'ekle',
    'aktar',
    'cek',
    'olustur',
    'bul',
    'ara',
    'goster',
    'listele',
    'oner',
    'duzenle',
    'degistir',
    'yayinla',
    'yayina',
    'al',
    'bana',
    'bir',
  };
  return text
      .split(' ')
      .where((token) => token.length > 1 && !stop.contains(token))
      .join(' ')
      .trim();
}

/// Blog domain yalnız açık içerik komutlarıyla seçilir.
/// "blog başlığını değiştir" gibi 46-alan storefront komutları burada
/// yakalanmaz.
VixrexAssistantDomainDecision routeVixrexAssistantDomain(String input) {
  final text = _normalize(input);
  if (text.isEmpty) {
    return const VixrexAssistantDomainDecision(
      domain: VixrexAssistantDomain.storefront,
    );
  }

  final mentionsBlog = text.contains('blog');
  final mentionsContent = _hasAny(text, ['yazi', 'makale', 'rehber', 'taslak']);
  if (!mentionsBlog || !mentionsContent) {
    return const VixrexAssistantDomainDecision(
      domain: VixrexAssistantDomain.storefront,
    );
  }

  final matches = <VixrexBlogIntent>[];
  if (
      text.contains('taslak') &&
      _hasAny(text, ['listele', 'goster', 'neler', 'hangileri'])) {
    matches.add(VixrexBlogIntent.blogTaslaklariniListele);
  }
  if (_hasAny(text, ['yayinla', 'yayina al'])) {
    matches.add(VixrexBlogIntent.blogYayinla);
  }
  if (
      _hasAny(text, ['duzenle', 'degistir']) &&
      _hasAny(text, ['yazi', 'taslak', 'makale'])) {
    matches.add(VixrexBlogIntent.blogTaslakDuzenle);
  }
  if (
      _hasAny(text, ['ekle', 'aktar', 'cek', 'taslak olustur', 'taslak ekle']) &&
      _hasAny(text, ['yazi', 'makale', 'rehber'])) {
    matches.add(VixrexBlogIntent.blogYazisiTaslakEkle);
  }
  if (text.contains('oner')) {
    matches.add(VixrexBlogIntent.blogYazisiOner);
  }
  if (_hasAny(text, ['bul', 'ara'])) {
    matches.add(VixrexBlogIntent.blogYazisiBul);
  }

  final unique = matches.toSet().toList();
  final query = _queryFrom(text);
  if (unique.length > 1) {
    return VixrexAssistantDomainDecision(
      domain: VixrexAssistantDomain.ambiguous,
      query: query,
    );
  }
  if (unique.length == 1) {
    return VixrexAssistantDomainDecision(
      domain: VixrexAssistantDomain.blog,
      intent: unique.first,
      query: query,
    );
  }
  return const VixrexAssistantDomainDecision(
    domain: VixrexAssistantDomain.storefront,
  );
}

String vixrexBlogIntentWireName(VixrexBlogIntent intent) {
  switch (intent) {
    case VixrexBlogIntent.blogYazisiBul:
      return 'blog_yazisi_bul';
    case VixrexBlogIntent.blogYazisiOner:
      return 'blog_yazisi_oner';
    case VixrexBlogIntent.blogYazisiTaslakEkle:
      return 'blog_yazisi_taslak_ekle';
    case VixrexBlogIntent.blogTaslaklariniListele:
      return 'blog_taslaklarini_listele';
    case VixrexBlogIntent.blogTaslakDuzenle:
      return 'blog_taslak_duzenle';
    case VixrexBlogIntent.blogYayinla:
      return 'blog_yayinla';
  }
}
